### ISSUE TITLE
Renames to avoid users think that our VMDBMS binding makes it less of a graph and relational engine

### ISSUE THREAD

**Issue by Starcounter-Jack** [Starcounter/Starcounter#2525]
Maybe we should avoid the confusion of Starcounter as simply being an object oriented database. Our VMDBMS binding does such a good job that some believe Starcounter an engine for serialising C# objects (as opposed to an interface to our graph/relational database).

Suggestions?

**Comment by Starcounter-Jack** [comment]
Key and KeyString?

**Comment by k-rus** [comment]
Starcounter will be counted as graph engine, when it will support graph-oriented select queries. Currently we can clarify that it is not just OO, but it is a real relational engine.

To the question. I don't actually understand it. What do you suggest to rename?

**Comment by Starcounter-Jack** [comment]
@k-rus I edited the initial issue post to reflect the naming issues.

**Comment by Starcounter-Jack** [comment]
The database models are graph models. Although users must traverse graphs (as in other graph databases that used to lack a query language), the reference model and the ```Relation``` class (heavily used by our clients) together with the emphasis on node traversing speed (due to in memory and VMDBMS) makes it suitable for representing huge graphs. I agree that path expressions and reference traversing does not mean that the query language is not needed, but the engine itself is what it is (even though it is lacking in the query department).

**Comment by Starcounter-Jack** [comment]
@k-rus Thanks for the observation. I edited the initial issue post to reflect the naming issues I intended to address.

**Comment by k-rus** [comment]
In my opinion the names are good enough and these properties are actually advanced features. I believe the confusion comes on earlier phase and not related to `ObjectNo` and `ObjectID`.

**Comment by per-samuelsson** [comment]
I think it's a good suggestion to rename ObjectId/ObjectNo, given the argumentation in the initial post.

> Key?
> Ref?
> EntityNo?

I strongly would vote for ```Key``` from the above three. Possibly the best alternative over all. ```Key``` for the 64-bit number, and ```KeyString``` for the string variant.

**Comment by Starcounter-Jack** [comment]
@per-samuelsson
I second that vote.

**Comment by k-rus** [comment]
I think until there is a specification for OID/ObjectId/ObjectNo/Key, which is under renaming consideration in this issue, it is bad idea to name it. OID is control by us, more specifically by kernel and without specification it is unclear how applications can use it. It is necessary to understand what happens with OIDs in different cases:

1. Stop/Start applications
2. Stop/Start database
3. Application migration, i.e., update of application schema
4. Starcounter upgrade
5. Create/Delete a database object/instance
6. Auto-generated dynamic types

How OIDs are preserved, are they reused?

We recently had a discussion how OIDs can be used in applications, especially in client cache, and we don't have answer to these questions, since there is no specification.

**Comment by Starcounter-Jack** [comment]
OIDs start at one and are never reused. We will allow to specify a starting sequence in future version. This is to aid master master replication setups.

**Comment by k-rus** [comment]
> OIDs start at one and are never reused. They are created on instantiation. This means that they are always set. They are incremented by one for each new tuple. They are unsigned 64 bit integers. This meabs that you can create a loop and keep creating tuples for thousands of years for millions of terabytes without running out of keys. We have a certain range reserved for system use that we should perhaps document, but otherwise there is not terribly much more to specify.

Thank you for the clarification! We should document this for our users.

The main disadvantage with using word `Key` for naming is that our key is the key and users keys are another keys and they should use another name.

**Comment by per-samuelsson** [comment]
> Thank you for the clarification! We should document this for our users.

It has worked like this since basically version 0.0.0.1. Only the introduction of the "system range" is a feature added later on. I don't think we need to document an implementation detail to our users, or at least its not a prio. The only thing they need to know is that identities are assigned by Starcounter, that they are unique for the lifetime of the database, and can not be altered.

> The main disadvantage with using word Key for naming is that our key is the key and users keys are another keys and they should use another name.

Doesn't this go for any kind of naming? I mean, ID suffering from that too. I still consider key the best.

**Comment by k-rus** [comment]
> The only thing they need to know is that identities are assigned by Starcounter, that they are unique for the lifetime of the database, and can not be altered.

There is few more things, which are important:

1. OID is never reused. I guess this is one of the meaning `identifiers cannot be altered`.
2. Do we preserve OIDs during database migration between different versions of Starcounter? This might be tricky until the type system implementation is completely settled.
3. What happens with OIDs during database migration between different versions of applications? I guess with new type system Starcounter should be able to preserve them.

I guess our goal is to guarantee that OIDs are preserved during life time of application database including migrations between different versions. However, it is user responsibility if the user wants to merge two databases in one, e.g., using reload functionality.



> Doesn't this go for any kind of naming? I mean, ID suffering from that too. I still consider key the best.

`ID` was not considered in the above discussion.

**Comment by Starcounter-Jack** [comment]
The OIDs are preserved between versions of Starcounter. They should be part of the unload and reload data.

**Comment by dan31** [comment]
@Starcounter-Jack, @k-rus emphasises the case when you want to use OID as an universally unique object identifier (aka UUID). Consider you rely on it in a hyperlink and save it in a browser. After you upgrade/merge the database, this OID might be no longer valid in the link. This means that we prohibit usage if OID as an UUID/token, true/false?

**Comment by Starcounter-Jack** [comment]
It would be a good idea to make the OID into a UUID (aka UID). But UIDs are 128bit. We cannot afford to use 128 bits inside the database. We would then have to salt the OID when showing it to the host. I.e. we keep OIDs internally and salt them to UIDs externally.  Any idea how this could be done? Increasing OID internal representation from 64 bit to 128 bit is of the table as the cost of performance would be extremely high. Size of database, cost of operation and size of indices would suffer as the CPU is 64 bit.

**Comment by dan31** [comment]
@Starcounter-Jack, you are absolutely right. Moreover, if UUID (even if 64 bit, however 64 isn't enough for having them not collided) is used instead of OID, it has to be generated per each object, which is much more complex than incrementing a counter. If we would have had clocks of infinite resolution, this might have helped to generate good OIDs :) One might consider some salted datetimestamp for this, though it appears that app developer should maintain UUIDs/tokens for the objects he/she needs independently. Using OID as a token we have seen as a common mistake, because it is not saved at migration, so there should be a hint on that matter on a page about OID.

**Comment by dan31** [comment]
However, see http://en.wikipedia.org/wiki/Globally_unique_identifier#Sequential_algorithms. Might make sense actually. Probably not difficult to take some implementation and check how it affects performance in common scenarios.

**Comment by Starcounter-Jack** [comment]
@per-samuelsson 

The new names: ```public ulong Key``` and ```public string KeyString```.

**Comment by per-samuelsson** [comment]
See also related issue #2532.

**Comment by k-rus** [comment]
> To merge two databases, you have to add a large delta number to one of them upon importing. I.e. you specify a number to add to all OIDs from the database you are importing. In this way, you can avoid collisions while maintaining foreign keys to OID primary keys on the imported data. This functionality is currently not available, so you would have to modify the reload text files yourself to merge databases.

The functionality is already available in the API of `Db.Unload`, but not in `staradmin`. 

> @Starcounter-Jack, @k-rus emphasises the case when you want to use OID as an universally unique object identifier (aka UUID). Consider you rely on it in a hyperlink and save it in a browser. After you upgrade/merge the database, this OID might be no longer valid in the link.

@dan31 thank you for clarifying.



> The new names: public ulong Key and public string KeyString.

I think this name (`Key`) brings the same limitation as keywords `When`, `Order`, etc., which we try avoid as much as possible. I think `Key` should not be reserved and be available for user applications.

**Comment by k-rus** [comment]
> @per-samuelsson
> 
> The new names: public ulong Key and public string KeyString.

I discussed this issue with @Starcounter-Jack. We cannot reserve `Key` and `Type` in public API, since very likely user wants to use property `Key` for setting some public key. Better names have to be found.
